src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
+                backtrack_stack.push(BacktrackFrame {
+                    context_backup: cx.clone(),
+                    deps_backup: remaining_deps.clone(),
+                    remaining_candidates: remaining_candidates,
+                    parent: parent.clone(),
+                    cur: cur,
+                    dep: dep.clone(),
+                    prev_active: prev_active,
+                    all_candidates: candidates,
+                });
+                candidate
+            }
+            None => {
+                // This dependency has no valid candidate. Backtrack until we
+                // find a dependency that does have a candidate to try, and try
+                // to activate that one.  This resets the `remaining_deps` to
+                // their state at the found level of the `backtrack_stack`.
+                trace!("{}[{}]>{} -- None", parent.name(), cur, dep.name());
+                let last_err = activation_error(&cx, registry, None, &parent, &dep,
+                                                &prev_active, &candidates);
+                try!(find_candidate(&mut backtrack_stack, &mut cx, &mut remaining_deps,
+                                    registry, last_err))
+            }
+        };
 
-    // Alright, for each candidate that's gotten this far, it meets the
-    // following requirements:
-    //
-    // 1. The version matches the dependency requirement listed for this
-    //    package
-    // 2. There are no activated versions for this package which are
-    //    semver-compatible, or there's an activated version which is
-    //    precisely equal to `candidate`.
-    //
-    // This means that we're going to attempt to activate each candidate in
-    // turn. We could possibly fail to activate each candidate, so we try
-    // each one in turn.
-    let mut last_err = None;
-    for candidate in my_candidates {
         trace!("{}[{}]>{} trying {}", parent.name(), cur, dep.name(),
                candidate.version());
-        let mut my_cx = cx.clone();
-        my_cx.resolve.graph.link(parent.package_id().clone(),
-                                 candidate.package_id().clone());
+        cx.resolve.graph.link(parent.package_id().clone(),
+                              candidate.package_id().clone());
 
         // If we hit an intransitive dependency then clear out the visitation
         // list as we can't induce a cycle through transitive dependencies.
         if !dep.is_transitive() {
-            my_cx.visited.clear();
-        }
-        let my_cx = try!(activate(my_cx, registry, candidate, &method,
-                                  &mut |cx, registry| {
-            activate_deps(cx, registry, parent, platform, deps.clone(), cur + 1,
-                          finished)
-        }));
-        match my_cx {
-            Ok(cx) => return Ok(Ok(cx)),
-            Err(e) => { last_err = Some(e); }
+            cx.visited.clear();
```

Comment:
This looks like a pre-existing bug, where the `visited` set stays cleared even after this branch of the deps tree has finished activating. I'd like to avoid fixing that until the main refactoring is done. :)